src/fields.jl

```diff
@@ -13,152 +13,111 @@ A type describing the location at the face of a grid cell.
+Used for the x-velocity field.
 """
-FaceFieldX(T, arch, grid) = Field(Face, Cell, Cell, zeros(T, arch, grid), grid)
+FaceFieldX(T, arch, grid) = Field{Face, Cell, Cell}(zeros(T, arch, grid), grid)
 
 """
     FaceFieldY([T=eltype(grid)], arch, grid)
 
-Return a `FaceFieldY` on architecture `arch` and `grid`.
+Return a `Field{Cell, Face, Cell}` on architecture `arch` and `grid`.
+Used for the y-velocity field.
 """
-FaceFieldY(T, arch, grid) = Field(Cell, Face, Cell, zeros(T, arch, grid), grid)
+FaceFieldY(T, arch, grid) = Field{Cell, Face, Cell}(zeros(T, arch, grid), grid)
 
 """
     FaceFieldZ([T=eltype(grid)], arch, grid)
 
-Return a `FaceFieldZ` on architecture `arch` and `grid`.
+Return a `Field{Cell, Cell, Face}` on architecture `arch` and `grid`.
+Used for the z-velocity field.
 """
-FaceFieldZ(T, arch, grid) = Field(Cell, Cell, Face, zeros(T, arch, grid), grid)
+FaceFieldZ(T, arch, grid) = Field{Cell, Cell, Face}(zeros(T, arch, grid), grid)
 
- CellField(arch, grid) = Field(Cell, Cell, Cell, arch, grid)
-FaceFieldX(arch, grid) = Field(Face, Cell, Cell, arch, grid)
-FaceFieldY(arch, grid) = Field(Cell, Face, Cell, arch, grid)
-FaceFieldZ(arch, grid) = Field(Cell, Cell, Face, arch, grid)
+ CellField(arch, grid) = Field((Cell, Cell, Cell), arch, grid)
+FaceFieldX(arch, grid) = Field((Face, Cell, Cell), arch, grid)
+FaceFieldY(arch, grid) = Field((Cell, Face, Cell), arch, grid)
+FaceFieldZ(arch, grid) = Field((Cell, Cell, Face), arch, grid)
 
-fieldtype(f::AbstractField) = typeof(f).name.wrapper
-location(::Field{Lx, Ly, Lz}) where {Lx, Ly, Lz} = (Lx, Ly, Lz)
+location(a) = nothing
+location(::AbstractLocatedField{X, Y, Z}) where {X, Y, Z} = (X, Y, Z)
+
+architecture(f::Field) = architecture(f.data)
+architecture(o::OffsetArray) = architecture(o.parent)
 
 @inline size(f::AbstractField) = size(f.grid)
-@inline length(f::AbstractField) = length(f.data)
+@inline length(f::Field) = length(f.data)
+
+@propagate_inbounds getindex(f::Field, inds...) = @inbounds getindex(f.data, inds...)
+@propagate_inbounds setindex!(f::Field, v, inds...) = @inbounds setindex!(f.data, v, inds...)
+@inline lastindex(f::Field) = lastindex(f.data)
+@inline lastindex(f::Field, dim) = lastindex(f.data, dim)
 
-@propagate_inbounds getindex(f::AbstractField, inds...) = getindex(f.data, inds...)
-@propagate_inbounds setindex!(f::AbstractField, v, inds...) = setindex!(f.data, v, inds...)
-@inline lastindex(f::AbstractField) = lastindex(f.data)
-@inline lastindex(f::AbstractField, dim) = lastindex(f.data, dim)
+"Returns `f.data` for `f::Field` or `f` for `f::AbstractArray."
+@inline data(a) = a
+@inline data(f::Field) = f.data
+
+"Returns `f.data.parent` for `f::Field`."
+@inline Base.parent(f::Field) = f.data.parent
 
 "Returns a view over the interior points of the `field.data`."
-@inline data(f::AbstractField) = view(f.data, 1:f.grid.Nx, 1:f.grid.Ny, 1:f.grid.Nz)
+@inline interior(f::Field) = view(f.data, 1:f.grid.Nx, 1:f.grid.Ny, 1:f.grid.Nz)
```

Comment:
It's more mundane than that. The function `data` now returns `field.data` for `field::Field`, and the object itself for an `AbstractOperation`. We use this because while `AbstractOperation` can be used on the GPU, a `Field` cannot and so we need to extract the underlying `OffsetArray` from it.

Reply:
`interior` returns the grid points in the model "interior" (ie, excluding halos). 

What would be more intuitive?

Hopefully in the future we can use dispatch on `Colon` (see #458) so that the user can ignore these functions altogether. I think we should strive for syntax that allows the user to think of a `Field` in the way one would think of a mathematical continuous field, rather than having to interact with the underlying discrete data and such.

Reply:
> The function data now returns field.data for field::Field, and the object itself for an AbstractOperation

I would prefer `values` or `data`. If `data` works, why are you using `interior`? Does `data` return halos as well? I was just curious.

Reply:
`values` is a julia `Base` function: for `Dict`-like objects it returns the values (see also the function `keys`)...

Maybe `interior_values`?

For `field::Field`, `data(field)` returns `field.data`, which is an `OffsetArray`. `interior` returns a *view* over the same offset array, which excludes the halo points. The difference only matters for broadcasting an operation over an array. For example, if you have an array of values you would like to set the solution to, then

```julia
field.data .= user_array
```

will fail, unless `user_array` includes halo points. This is undesirable because halos are arbitrary. Thus they would use

```julia
interior(field) .= user_array
```
which due to the view, will work when `user_array` contains only interior values. 

However, I think we should encourage users to use our functions rather than manually setting values, as in:

```julia
set!(field, user_array)
```

For plotting `interior` would also prove useful if you don't want to plot halos. Almost always, however, we are plotting slices (not the whole array); in that case dispatch on `Colon` may suffice since we can design `getindex` such that

```julia
field[:, 1, :]
```
returns a slice over interior values, excluding halos.